subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java

```diff
@@ -175,9 +177,29 @@ public File getSingleFile() throws IllegalStateException {
 
     @Override
     public String getAsPath() {
+        showGetAsPathDeprecationWarning();
         return GUtil.asPath(this);
     }
 
+    private void showGetAsPathDeprecationWarning() {
+        List<String> filesAsPaths = this.getFiles().stream()
+            .map(File::getPath)
+            .filter(path -> path.contains(File.pathSeparator))
+            .collect(Collectors.toList());
+        if (!filesAsPaths.isEmpty()) {
+            DeprecationLogger.deprecateBehaviour(String.format(
+                    "FileCollection '%s' that is being mapped to a path contains already concatenated files instead of a collection of single files." +
+                        " This can lead to uncontrolled failures. Problematic concatenations are: %s.",
+                    getDisplayName(),
+                    String.join(", ", filesAsPaths))
+                )
+                .withAdvice("Add files to a collection as single files instead of manually concatenating them.")
+                .willBecomeAnErrorInGradle8()
+                .undocumented()
```

Comment:
This needs to be documented in the upgrading notes.

Reply:
fixed